device/devicegate/filter.go

```diff
@@ -0,0 +1,212 @@
+	DeleteFilter(key string) bool
+
+	// GetAllowedFilters returns the set of filters that devices are allowed to be filtered by. Also returns a
+	// bool that is true if there are allowed filters set, and false if there aren't (meaning that all filters are allowed)
+	GetAllowedFilters() (Set, bool)
+}
+
+// Set is an interface that represents a read-only hashset
+type Set interface {
+	json.Marshaler
+	// Has returns true if a value exists in the set, false if it doesn't.
+	Has(interface{}) bool
+
+	// VisitAll applies the visitor function to every value in the set.
+	VisitAll(func(interface{}))
+}
+
+// FilterStore can be used to store filters in the Interface
+type FilterStore map[string]Set
+
+// FilterSet is a concrete type that implements the Set interface
+type FilterSet map[interface{}]bool
+
+// FilterGate is a concrete implementation of the Interface
+type FilterGate struct {
+	FilterStore    FilterStore `json:"filters"`
+	AllowedFilters Set         `json:"allowedFilters"`
+
+	lock sync.RWMutex
+}
+
+type FilterRequest struct {
+	Key    string        `json:"key"`
+	Values []interface{} `json:"values"`
+}
+
+func (f *FilterGate) VisitAll(visit func(string, Set) bool) int {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	visited := 0
+	for key, filterValues := range f.FilterStore {
+		visited++
+		if !visit(key, filterValues) {
+			break
+		}
+	}
+
+	return visited
+}
+
+func (f *FilterGate) GetFilter(key string) (Set, bool) {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	v, ok := f.FilterStore[key]
+	return v, ok
+
+}
+
+func (f *FilterGate) SetFilter(key string, values []interface{}) (Set, bool) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	oldValues := f.FilterStore[key]
+	newValues := make(FilterSet)
+
+	for _, v := range values {
+		newValues[v] = true
+	}
+
+	f.FilterStore[key] = newValues
+
+	if oldValues == nil {
+		return oldValues, true
+	}
+
+	return oldValues, false
+
+}
+
+func (f *FilterGate) DeleteFilter(key string) bool {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	_, ok := f.FilterStore[key]
+
+	if ok {
+		delete(f.FilterStore, key)
+		return true
+	}
+
+	return false
+}
+
+func (f *FilterGate) AllowConnection(d device.Interface) (bool, device.MatchResult) {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	for filterKey, filterValues := range f.FilterStore {
+		// check for filter match
+		if found, result := f.FilterStore.metadataMatch(filterKey, filterValues, d.Metadata()); found {
+			return false, result
+		}
+	}
+
+	return true, device.MatchResult{}
+}
+
+func (f *FilterGate) GetAllowedFilters() (Set, bool) {
+	if f.AllowedFilters == nil {
+		return f.AllowedFilters, false
+	}
+
+	return f.AllowedFilters, true
+}
+
+func (s FilterSet) Has(key interface{}) bool {
+	return s[key]
+}
+
+func (s FilterSet) VisitAll(f func(interface{})) {
+	for key := range s {
+		f(key)
+	}
+}
+
+func (s FilterSet) MarshalJSON() ([]byte, error) {
+	temp := make([]interface{}, len(s))
```

Comment:
VisitAll holds a lock, but this method does not.  That means that elements of the slice can be updated concurrently to this method.  Is that ok?  Consider this scenario:

```
var s FilterSet // initialize somehow
myValue := new(int)
s["key"] = myValue
go s.MarshalJSON()
go func() {
    *myValue = 123 // an update that may occur simultaneously with MarshalJSON
}()
```

I can't tell if that's an issue, however.  Do we anticipate modifying values *after* they are put into a `FilterSet`?  If so, then `MarshalJSON` and any other mutating method will need to hold a lock.

Reply:
I intended for FilterSet to implement the Set interface, which is supposed to be a read-only Set data structure. So the intention was for FilterSet to be read-only. But if someone were to use FilterSet by itself and not as a Set interface, they certainly could modify values after they put it into the FilterSet, so it would probably be safer to include a lock here